File: PPA_PixiJS-Angular_Prototype/app/js/canvas/graph/graph.js
/**
 * Created by mdaly on 12/31/13.
 */

Graph = function(width, height) {
    this.ChannelCount = 1;
    this.HorizontalDivisions = 500; // number of quads, left to right

    this.xAxis = new Graph.AxialInfo();
    this.yAxis = new Graph.AxialInfo();

    this.yAxisLabels = {}; // label name to label object
    this.xAxisLabels = {}; // label name to label object

    this.xLabelOffset = 10;
    this.yLabelOffset = 10;

    this.context = new RenderContext(width, height, this.onReady, this);

    this.graphics = new PIXI.Graphics;
    this.context.AddObject(this.graphics);
    this.drawAxes = false;

    var me = this;
    requestAnimationFrame(this.Update.bind(this));
}

Graph.GraphMode = {
    Linear: 0,
    Exponential: 1
}

Graph.AxialInfo =  function() {
    this.min        = 0;
    this.max        = 0;
    this.graphMode  = Graph.GraphMode.Linear;
    this.subminor   = new Array();
    this.minor      = new Array();
    this.major      = new Array();
    this.tick       = new Array();
    this.label      = new Array();
}

Graph.prototype.Init = function() {

}

/**
 * Converts from axis-space to control-space
 * @param graphX
 * @returns {number}
 */

Graph.prototype.graphToNodeX = function (graphX) {
    var norm,
        me = this;

    if (me.xAxis.mode == Graph.GraphMode.Linear) {
        norm = (graphX - me.xAxis.min) / (me.xAxis.max - me.xAxis.min);
    }
    else {
        var expRange =  BA.Utils.Log10(me.xAxis.max / me.xAxis.min);

        norm = BA.Utils.Log10(graphX / me.xAxis.min) / expRange;
    }

    return norm * me.MaxX();
}

/**
 * Converts from axis-space to control-space
 * @param graphX
 * @returns {number}
 */
Graph.prototype.graphToNodeY = function(graphY) {
    var norm,
        me = this;

    if (me.yAxis.mode == Graph.GraphMode.Linear) {
        norm = (graphY - me.yAxis.min) / (me.yAxis.max - me.yAxis.min);
    }
    else {
        var expRange = BA.Utils.Log10(me.yAxis.max / me.yAxis.min);

        norm = BA.Utils.Log10(graphY / me.yAxis.min) / expRange;
    }

    return norm * (me.MaxY());
}

/**
 * Converts from control-space to axis-space
 * @param graphX
 * @returns {number}
 */
Graph.prototype.nodeToGraphX = function(nodeX) {
    var me = this,
        norm = (nodeX) / (me.context.Width());

    if (me.xAxis.mode == Graph.GraphMode.Linear) {
        return norm * (me.xAxis.max - me.xAxis.min) + me.xAxis.min;
    }
    else {
        var expRange = BA.Utils.Log10(me.xAxis.max / me.xAxis.min);

        return Math.pow(10.0, norm * expRange) * me.xAxis.min;
    }
}

/**
 * Converts from control-space to axis-space
 * @param graphX
 * @returns {number}
 */
Graph.prototype.nodeToGraphY = function(nodeY) {
    var me = this,
        norm = (nodeY) / (me.context.Height());

    if (me.yAxis.mode == Graph.GraphMode.Linear) {
        return norm * (me.yAxis.max - me.yAxis.min) + me.yAxis.min;
    }
    else {
        var expRange = BA.Utils.Log10(me.yAxis.max / me.yAxis.min);

        return Math.pow(10.0, norm * expRange) * me.yAxis.min;
    }
}

/**
 * Converts from axis-space to control-space
 * @param graphX
 * @returns {number}
 */
Graph.prototype.graphToNode = function(graph) {
    return new PIXI.Point(this.graphToNodeX(graph.x), this.graphToNodeY(graph.y));
}

/**
 * Converts from control-space to axis-space
 * @param graphX
 * @returns {number}
 */
Graph.prototype.nodeToGraph = function(node) {
    return new PIXI.Point(this.nodeToGraphX(node.x), this.nodeToGraphY(node.y));
}

Graph.prototype._drawLine = function(p1, p2, width) {
    this.graphics.moveTo(p1.x, p1.y);
    this.graphics.lineTo(p2.x, p2.y);
}

/**
 * returns x minimum in control-space
 * @returns {number}
 * @constructor
 */
Graph.prototype.MinX  = function() {
    return 0.0;
}

/**
 * returns x minimum in control-space
 * @returns {number}
 * @constructor
 */
Graph.prototype.MinY = function() {
    return 0.0;
}

/**
 * Returns y maximum in control-space
 * @returns {*}
 * @constructor
 */
Graph.prototype.MaxY  = function() {
    return this.context.Height();
}

/**
 * Returns x maximum in control-space
 * @returns {*}
 * @constructor
 */
Graph.prototype.MaxX = function() {
     return this.context.Width();
}

Graph.prototype.Width = function() {
    return this.context.Width();
}

Graph.prototype.Height = function() {
    return this.context.Height();
}

Graph.prototype.DrawXAxis = function() {
    var majors = this.xAxis.major,
        y1 = this.MaxY(),
        y2 = this.MinY();
    for (var i = 0; i < majors.length; i++) {
        var x = this.graphToNodeX(majors[i]);

        this._drawLine(new PIXI.Point(x , y1), new PIXI.Point(x, y2), 3.0);
    }

    var minors = this.xAxis.minor;
    for (var i = 0; i < minors.length; i++) {
        var x = this.graphToNodeX(minors[i]);

        this._drawLine(new PIXI.Point(x,y1), new PIXI.Point(x, y2), 1.0);
    }
}

Graph.prototype.DrawYAxis = function() {
    var majors = this.yAxis.major,
        x1 = this.MinX(),
        x2 = this.MaxX();

    // Y AXIS FEATURES
    for (var i = 0; i < majors.length; i++) {
        var y = this.graphToNodeY(majors[i]);

        this._drawLine(new PIXI.Point(x1, y), new PIXI.Point(x2, y), 3);
    }

    var minors = this.yAxis.minor;
    for (var i = 0; i < minors.length; i++) {
        var y = this.graphToNodeY(minors[i]);

        this._drawLine(new PIXI.Point(x1, y), new PIXI.Point(x2, y), 1);
    }
}

Graph.prototype.Update = function() {
    var me = this;

    if (me.context.requestRenderFrame) {

        if (me.drawAxes) {
            me.graphics.clear();
            me.graphics.lineStyle (1.5, 0x1F5C20, 1);
            me.DrawXAxis();
            me.DrawYAxis();
        }

        me.context.Update();
        me.context.requestRenderFrame = false;
    }

    //Keep asking for updates
    requestAnimationFrame(function() { me.Update(); } );
}

Graph.prototype.resize = function() { this.context.resize(); }
Graph.prototype.GetDomElement = function() { return this.context.GetDomElement(); }

File: PPA_QT_Prototype/src/cpp/graph.cpp
/**
 * Graph is a base class for anything that needs to render axial details
 * It handles drawing of vertial and horizontal lines, ticks, labels
 *
 */

#include "graph.h"

#include <stdio.h>
#include <math.h>

#include <QtQuick/qsgflatcolormaterial.h>
#include <QSGSimpleRectNode>
#include <QVariant>
#include <QQmlProperty>
#include <QDate>
#include <qsggeometry.h>
#include <QQmlEngine>
#include <QQmlContext>
#include <QDir>

#include "Constants.h"

Graph::Graph(QQuickItem *parent) : QQuickItem(parent)
{
    Q_INIT_RESOURCE(resources);

    setFlag(ItemHasContents, true);

    xLabelOffset = 0;
    yLabelOffset = 0;

    labelComponent = NULL;

    //HorizontalDivisions = 500;
}

QSGNode* Graph::updatePaintNode(QSGNode *oldNode, UpdatePaintNodeData *data)
{   
    QSGNode* rootNode = QQuickItem::updatePaintNode(oldNode, data);

    if (rootNode == NULL) {
        rootNode = new QSGNode();
        rootNode->setFlag(QSGNode::OwnedByParent, true);
    }
    else {
        rootNode->removeAllChildNodes();
    }

    DrawAxes(rootNode);

    return rootNode;
}

void Graph::DrawAxes(QSGNode * rootNode) {
    if (!this->m_simple) {
        //DrawBackground(rootNode);
        DrawXAxis(rootNode);
        DrawYAxis(rootNode);

    }
}

void Graph::DrawBackground(QSGNode * rootNode) {
    QSGSimpleRectNode * node = new QSGSimpleRectNode(boundingRect(), m_color);
    rootNode->appendChildNode(node);
}

void Graph::DrawXAxis(QSGNode * rootNode) {
    QRectF bounds = boundingRect();
    // X AXIS FEATURES
    for (QList<float>::iterator i = xAxis.major.begin(); i != xAxis.major.end(); i++) {
        float x = graphToNodeX(*i);
        float y = boundingRect().height() + xLabelOffset;

        _drawLine(rootNode, m_primaryLineColor, x, bounds.top(), x, bounds.bottom(), 2);
        _drawLabel(x, y, QString::number(*i), xAxisLabels);
    }

    for (QList<float>::iterator i = xAxis.minor.begin(); i != xAxis.minor.end(); i++) {
        float x = graphToNodeX(*i);
        float y = boundingRect().height() + xLabelOffset;

        _drawLine(rootNode, m_primaryLineColor.darker(),x, bounds.top(), x, bounds.bottom(), -1);
        _drawLabel(x, y, QString::number(*i), xAxisLabels);
    }

    for (QList<float>::iterator i = xAxis.subminor.begin(); i != xAxis.subminor.end(); i++) {
        float x = graphToNodeX(*i);
        float y = boundingRect().height() + xLabelOffset;

        _drawLine(rootNode, m_primaryLineColor.darker(), x, bounds.bottom(), x, bounds.top(), 0.5);
        _drawLabel(x, y, QString::number(*i), xAxisLabels);
    }

    // We can draw ticks, but the PPA prototype didn't have any
    /*for (QList<float>::iterator i = xAxis.tick.begin(); i != xAxis.tick.end(); i++) {
        float x = graphToNodeX(*i);

        _drawLine(rootNode, QColor(255, 255, 255, 255), x, bounds.bottom(), x, bounds.bottom()+5, 2);
    }*/
}

void Graph::DrawYAxis(QSGNode * rootNode) {
    QRectF bounds = boundingRect();
    // Y AXIS FEATURES
    for (QList<float>::iterator i = yAxis.major.begin(); i != yAxis.major.end(); i++) {
        float x = yLabelOffset;
        float y = graphToNodeY(*i);

        _drawLine(rootNode, m_primaryLineColor, bounds.left(), y, bounds.right(), y, 2);
        _drawLabel(x, y, QString::number(*i), yAxisLabels);
    }

    for (QList<float>::iterator i = yAxis.minor.begin(); i != yAxis.minor.end(); i++) {
        float x = yLabelOffset;
        float y = graphToNodeY(*i);

        _drawLine(rootNode, m_primaryLineColor.darker(), bounds.left(), y, bounds.right(), y, 1);
        _drawLabel(x, y, QString::number(*i), yAxisLabels);
    }

    for (QList<float>::iterator i = yAxis.subminor.begin(); i != yAxis.subminor.end(); i++) {
        float x = yLabelOffset;
        float y = graphToNodeY(*i);

        _drawLine(rootNode, m_primaryLineColor.darker(), bounds.left(), y, bounds.right(), y, 1);
        _drawLabel(x, y, QString::number(*i), yAxisLabels);
    }

    // We can draw ticks, but the PPA prototype didn't have any
    /*for (QList<float>::iterator i = yAxis.tick.begin(); i != yAxis.tick.end(); i++) {
        float y = graphToNodeY(*i);

        _drawLine(rootNode, QColor(255, 255, 255, 255), bounds.left(), y, bounds.left()-5, y, 2);
    }*/
}

float Graph::graphToNodeX(float graphX) {
    float norm;

    if (xAxis.mode == Linear) {
        norm = (graphX - xAxis.min) / (xAxis.max - xAxis.min);
    }
    else {
        float expRange = log10f(xAxis.max / xAxis.min);

        norm = log10f(graphX / xAxis.min) / expRange;
    }

    return norm * (boundingRect().width());
}

float Graph::graphToNodeY(float graphY) {
    float norm;

    if (yAxis.mode == Linear) {
        norm = (graphY - yAxis.min) / (yAxis.max - yAxis.min);
    }
    else {
        float expRange = log10f(yAxis.max / yAxis.min);

        norm = log10f(graphY / yAxis.min) / expRange;
    }

    return norm * (boundingRect().height());
}

float Graph::nodeToGraphX(float nodeX) {
    float norm = (nodeX) / (boundingRect().width());

    if (xAxis.mode == Linear) {
        return norm * (xAxis.max - xAxis.min) + xAxis.min;
    }
    else {
        float expRange = log10f(xAxis.max / xAxis.min);

        return powf(10.0f, norm * expRange) * xAxis.min;
    }
}

float Graph::nodeToGraphY(float nodeY) {
    float norm = (nodeY) / (boundingRect().height());

    if (yAxis.mode == Linear) {
        return norm * (yAxis.max - yAxis.min) + yAxis.min;
    }
    else {
        float expRange = log10f(yAxis.max / yAxis.min);

        return powf(10.0f, norm * expRange) * yAxis.min;
    }
}

QVector2D Graph::graphToNode(QVector2D graph) {
    QVector2D ret;
    ret.setX(graphToNodeX(graph.x()));
    ret.setY(graphToNodeY(graph.y()));
    return ret;
}

QVector2D Graph::nodeToGraph(QVector2D node) {
    QVector2D ret;
    ret.setX(nodeToGraphX(node.x()));
    ret.setY(nodeToGraphY(node.y()));
    return ret;
}

void Graph::_drawLine(QSGNode * rootNode, QColor color, float startx, float starty,  float endx, float endy, float width) {
    QSGSimpleRectNode * node = new QSGSimpleRectNode();
    node->setFlag(QSGNode::OwnedByParent);
    node->markDirty(QSGNode::DirtyForceUpdate);

    if (startx == endx) { //Vertical Line
        node->setRect(startx, endy, -width, starty - endy);
    }
    else { //Horizontal Line
        node->setRect(startx, endy, endx - startx, width);
    }

    node->setColor(color);
    rootNode->appendChildNode(node);
}

void Graph::_drawLabel(float x, float y, QString text, QHash<QString, QQuickItem*> &labelCache) {
    if(labelComponent == NULL) {
        QQmlEngine * engine = new QQmlEngine();
        labelComponent = new QQmlComponent(engine, QUrl("qrc:/src/qml/controls/common/_Text.qml", QUrl::StrictMode));

        if( labelComponent->status() != QQmlComponent::Ready)
        {
            qDebug() << ("Error:"+ labelComponent->errorString() );
            return; // or maybe throw
        }
    }

    if( labelComponent->status() != QQmlComponent::Ready)
    {
        return;
    }


    QQuickItem * label = labelCache[text];

    if (!label) { //TODO: Someday, scene graph might have a better text renderer
        label= qobject_cast<QQuickItem*>(labelComponent->create());
        label->setParentItem(this->parentItem()); //TODO: Who REALLY owns this item? The component? Or the Hash?

        label->setProperty("text", text);
        label->setProperty("width", "parent.width");
        label->children().first()->setProperty("color", m_primaryLineColor.darker());

        labelCache[text] = label;
    }

    if (label != NULL) {
        label->setProperty("x", x);
        label->setProperty("y", y);
    }
}

